src/mlpack/methods/hdbscan/hdbscan_impl.hpp

```diff
@@ -0,0 +1,763 @@
+{
+  // Label all the clusters which can be root
+  sort(clusters.begin(), clusters.end());
+  std::map<size_t, size_t> labelOfClusters;
+  for (size_t i = 0; i < clusters.size(); i++)
+    labelOfClusters[clusters[i]] = i;
+
+  // Find the smallest cluster which can be the root
+  // Condensed tree contan all those clusters
+  // whose child cluster has size >= 1
+  // Minimum will be that cluster which has one point
+  // And will be equal to size of input original data set
+  size_t rootCluster = arma::min(condensedTree.row(0));
+  result.set_size(1, rootCluster);
+  emst::UnionFind unionTree(arma::max(condensedTree.row(0))+1);
+
+  // Go through all edges in the condensed tree
+  // If the child cluster of edge cannot be root of any cluster
+  // Merge those clusters
+  for (size_t i = 0; i < condensedTree.n_cols; i++)
+    if (!std::binary_search(clusters.begin(),
+                           clusters.end(),
+                           condensedTree(1, i)))
+      unionTree.Union(condensedTree(0, i), condensedTree(1, i));
+
+  // Now assign labels to all the points of the dataset
+  size_t currentPoint, parentOfCurrentPoint;
+
+  // If only one cluster is possible
+  // Check lambda value of all the children
+  // If lambda_p >= lambda_rootCluster
+  // p is included in cluster
+  // Otherwise it is a noise
+  if (clusters.size() == 1 && allowSingleCluster)
+  {
+    // find lambda of root cluster
+    double lambdaRootCluster = 0;
+    for (size_t i = 0; i < condensedTree.n_cols; i++)
+    {
+      if (condensedTree(0, i) == rootCluster)
+        lambdaRootCluster = std::max(lambdaRootCluster, condensedTree(2, i));
+    }
+
+    // Mark all the points as noises
+    for (size_t i = 0; i < rootCluster; i++)
+      result[i] = SIZE_MAX;
+
+    double eps = std::numeric_limits<double>::epsilon();
+    double eps_error = pow(condensedTree.n_cols, 2) * eps;
+    // consider all edges in condensed tree
+    // whose child is a single point (not cluster)
+    // check if their lambda has value
+    // less then that of the root cluster
+    for (size_t i = 0; i < condensedTree.n_cols; i++)
+    {      currentPoint = condensedTree(1, i);
+      if (currentPoint >= rootCluster)  continue;
+      parentOfCurrentPoint = unionTree.Find(currentPoint);
+      if (parentOfCurrentPoint == rootCluster &&
+        condensedTree(2, i) + eps_error >= lambdaRootCluster)
+        result[currentPoint] = 0;
+    }
+
+    return;
+  }
+
+  for (currentPoint = 0; currentPoint < rootCluster; currentPoint++)
+  {
+    parentOfCurrentPoint = unionTree.Find(currentPoint);
+    if (parentOfCurrentPoint <= rootCluster)
+      // label these points as noise
+      result[currentPoint] = SIZE_MAX;
+    else
+      // assign the correct cluster
+      result[currentPoint] = labelOfClusters[parentOfCurrentPoint];
+  }
+} // function getLabels ends
+
+/**
+ * This functin provides label to each and every point
+ * Noise points are labelled as ( total number of points + 1)
```

Comment:
The DBSCAN implementation labels these as SIZE_MAX; you could do the same here, or if you have a good reason for your choice we can change DBSCAN too.